### DISCUSSION TITLE
tutorials/physics/physics_introduction

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#2] [User-contributed notes]
# tutorials/physics/physics_introduction

In game development, you often need to know when two objects in the game intersect or come into contact. This is known as collision detection. When a collision is detected, you typically want somet...

https://docs.godotengine.org/en/latest/tutorials/physics/physics_introduction.html

**Comment by jerryjrowe** [comment]
Area2D will only detect StaticBody2D when calling get_overlapping_bodies() if the Area2D has monitorable set to true. You'd think it'd only need monitoring set to true, but it also needs to be monitorABLE for StaticBody2D.

**Reply by jh2952** [reply]
Thanks for this comment - helped me solve my issue :)

**Comment by miketvo** [comment]
The documentation is not clear about this:

Non-moving `Area` will not fire `body_entered` and `body_exited`  signals for moving `StaticBody` and frozen `RigidBody`, because those physics bodies are not intended to be moved. Use `AnimatableBody` instead if you want to move them and have them be detected by `Area`.

A *moving* `Area` however will detect non-moving `StaticBody` and frozen `RigidBody`.

See godotengine/godot#81883

**Reply by gargantuan** [reply]
Three hours I've wasted trying to figure this one out. Thank you!

**Reply by cucumberlush** [reply]
Makes more sense great Job figuring that one out

**Comment by gargantuan** [comment]
An easy way to remember the difference between collision layers and collisions masks that I find useful.

Collision Layer: I live here
Collision Mask: I will look here

**Reply by JuhaW** [reply]
Display/View Layer: This displays/Views objects
Collision Layer: Here collision happens 

:)

**Reply by thenicksager** [reply]
If the collision layer is "I live here", what kinds of objects could (or should) ever have multiple collision layers set? What kind of objects can "be" in more than one place at once?

**Reply by flowstate** [reply]
You're misunderstanding what "I live here" means. It's not referring to a physical location, it's referring to a physics layer. 

For a contrived example, let's say I have a layer for 'terrain' (floors, walls, etc) and another for 'hazards' (spikes, acid). If I want my player to be hurt by spikes, but to be able to walk on them while invincible or if they have a specific item, then my spikes could live in both 'terrain' and 'hazards'. 

This would allow my spikes to behave as a hazard AND be walkable like terrain. Does that make sense?

**Comment by v1993** [comment]
I feel like the example where `pow` is used to compute collision masks would be better off using bit shifts, at least since these are semantically bit operations on integers and `pow` works with floats. That would make it

```gdscript
(1 << 1 - 1) + (1 << 3 - 1) + (1 << 4 - 1)
```

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/10943

**Comment by mj2068** [comment]
That Asteroid example was not good, the code only calls apply_force() and apply_torque(), just through the `state` argument, didn't actually directly change any properties which it was mainly talking about prior to the snippet. These calls we can already use in _physics_process(). This didn't convey the main advantages of _integrate_forces(). It's confusing for people trying to understand the difference.

**Reply by backslash-enn** [reply]
Exactly, this was the only part that really threw me off.

To put it concisely, they say "only use `_PhysicsProcess` for abc, dont use it for xyz. If you want to do xyz, use `_IntegrateForces` instead". Sounds good

Then they say "here's an example of us using `_IntegrateForces`"....and in the example they only do abc

Then they double down in the footnote below that example and say "note how we're only doing abc here, not xyz"

...isnt the entire reason  `_IntegrateForces` exists so that we *can* do xyz? I think that section would benefit from a rewrite

**Reply by Calinou** [reply]
This is being tracked in https://github.com/godotengine/godot-docs/issues/10029.

**Comment by Vitorgus** [comment]
I think it would be nice to explain and include some reference to the [Joint2D](https://docs.godotengine.org/en/stable/classes/class_joint2d.html) nodes ([DampedSpringJoint2D](https://docs.godotengine.org/en/stable/classes/class_dampedspringjoint2d.html), [GrooveJoint2D](https://docs.godotengine.org/en/stable/classes/class_groovejoint2d.html) and [PinJoint2D](https://docs.godotengine.org/en/stable/classes/class_pinjoint2d.html)). Or even a separated page just talking about the 2D and 3D Joints, since they are related to the physics system. They are only mentioned in the documentation, but never really explained or shown examples of use.

**Comment by Bellbirdbits** [comment]
Currently move_and_collide() can be set to only test, but only returns one collision.
move_and_slide() automatically resolves position and can return multiple collisions.
test_collision() only returns a bool.
The result is, custom filtering is difficult and always requires a second object attached if you want a test-but-dont-stop (for picking up items for example), this makes it complicated if you have multiple collision shapes, changing collision shapes (a character that can crouch for example), collecting multiple objects in a single move

Consider a get_collisions() that works like move_and_collide() but reports multiple collisions like slide. Maybe it can have a custom mask as an argument?
Could area2d/3d be split in to CollisionArea which only has the overlapping_body functions and Area2d/3d which extends CollisionArea?
CollisionArea comes with a lot of extra data about gravity and audio, which largely goes unused.

Is there a way to have move_and_slide() report an object but not respond to it?
I'm often bumping in to this limitation and working around it, have I missed something?

**Reply by LoganWyborn** [reply]
Without more context it's hard to say for your specific situation, but based on the examples you've given, here's some ideas that might help.

Picking up items: Have the item be an Area which detects the entering body, while the body can ignore it (through layers/masking.) This way it doesn't interfere with your MoveAndCollide. 

Crouching: Have two CollisionShapes (or two Shape2D/3D resources) that are enabled/disabled or swapped out based on what state the character is in. 

 For something akin to MoveAndSlide which reports but doesn't automatically respond to collisions, Ray-casting/Shape-casting should work. If you need to get multiple collisions from the result, you can do something like: 
1. run your casting query
2. if result is not empty, store in some list
3. add result collider to query exclusion list
4. repeat until result comes back empty

This is a pretty simple structure that should give you all the desired data. It shouldn't be too cumbersome unless you are constantly checking a very large number of collisions. Though if this is the case, I suspect you'll have performance issues regardless.

**Comment by LoganWyborn** [comment]
A couple questions regarding the note about "Be careful to never scale your collision shapes in the editor.":

1. Does this only apply to the CollisionShape node itself? What if the parent Physics Body, or a parent further up the tree, is scaled, is this okay or does it introduce problems also? 
2. Is changing the scale in code safe, or should this be avoided as well?
3. Can anyone elaborate further on why exactly this is not recommended?

Some further info here would be greatly appreciated. While the note is only briefly discussed, it seems worth more focus. Depending on the answers to the above questions, I can imagine a situation where the use of Collision Shapes introduces design considerations/constraints on the entire project. I'd be very thankful to have more clarity on when and why this matters.

**Reply by Calinou** [reply]
Note that this text is a little outdated, and *uniform* collision shape scaling (i.e. same scale on all axes) should usually work OK, particularly when using Jolt Physics in 3D. Non-uniform scaling remains problematic though, and can lead to all sorts of physics issues, such as collisions not matching the visual representation of the debug shape.

> 1. Does this only apply to the CollisionShape node itself? What if the parent Physics Body, or a parent further up the tree, is scaled, is this okay or does it introduce problems also?

This applies to the CollisionShape and any parents of it.

> 2. Is changing the scale in code safe, or should this be avoided as well?

This should be avoided as well, as the physics issues will occur no matter how the scale was set.

> 3. Can anyone elaborate further on why exactly this is not recommended?

The physics engine (GodotPhysics) can't handle all situations of scaling, particularly non-uniform scaling. Uniform scaling will work somewhat OK, but non-uniform scaling or negative scaling won't.

In 3D, when using Jolt Physics, certain cases of non-uniform scaling can be handled correctly (e.g. a box where the X and Z axes' scale are identical, but the Y axis scale is different). This will not work for all shape types though. You'll see a warning being printed when such a situation is encountered.

**Reply by LoganWyborn** [reply]
Interesting, thanks for the quick reply. This does bring up some follow-on questions for me. Mostly, I am trying to envision the design pattern for a project that uses these physics tools and has elements, or perhaps entire sections, where scale might change (and whether consideration must be given to this very early in the project design process). These questions are quite related, so the answer might address multiple of them, but I'll separate them out again for organization:

1. If the *net* scaling of a node cancels out, is this still an issue? e.g. Grandparent has a scale of 2, Parent has a scale of 0.5, thus child's net scale is effectively 1. Is this okay, or does it still cause problems? 
2. Does this create a design constraint where any node which might ever have a Collision Shape anywhere in it's children can't/shouldn't be scaled (or at least, only uniformly scaled)? 
3. Is there a recommended design pattern that instead alleviates this issue, such as the node detecting when it's global_scale would be changed and modifying it's Shape2D resource instead? 
4. Does this mean that for a project which requires this functionality, we should be designing elements that are capable of "working around" this through some custom scaling behaviour which doesn't touch the Node2D scale property at all? 
5. (A more philosophical question that might not have a direct answer) If there is no "simple" solution to these problems, how does this fit into the loosely coupled/ideally context-independent design philosophy encouraged in Godot's best practices? 

To my understanding, this introduces either a project-wide design constraint, or every CollisionShape becomes context-dependent. This seems to conflict with the core design principles encouraged throughout this documentation, so while this specific case may be niche, understanding how to logically rectify this does seem quite important in my mind. 

Thanks again for the additional info, your help with understanding this is invaluable. Apologies if these questions are burdensome, I am hopeful these detailed clarifications can help avoid frustrating problems in the future, especially if this requires essentially ground-up consideration in the architecture of a project.

**Comment by p69549** [comment]
While this documentation has done a good job at explaining collisions, there should be an explanation of how the code works and what it does.

**Reply by Calinou** [reply]
Which part of the code are you referring to?

**Reply by p69549** [reply]
While I'm referring to the code for bounce(), I think it's important that any code included in these pages should have at least a small explanation of what they do, rather than just throwing code at the user.

Otherwise, this documentation ends up teaching new programmers that they should just copy code without understanding how it works.
